### PR TITLE
Prop equal against circular structures

### LINKED ISSUE #411: assert.propEqual not guarded against circular structures
As a result of #343 we got a new "propEqual" method which is usefully capable of comparing object trees in a "constructor-blind" manner, using a weaker semantic than QUnit.deepEqual. However, this method is not up to the standard of the rest of the framework - in particular, QUnit.deepEqual itself offers support for
1. correctly comparing cyclic structures without bombing the stack (#100 - recently improved  by #397)
2. not being confused by primitive arguments such as strings - these are compared properly and highlighted in the difference output.

"propEqual" was implemented using a utility function "objectValues" which performs a clone of the arguments before they are dispatched to QUnit.deepEqual, and it is deficiencies in this algorithm which are responsible for deficiencies in propEqual as compared to deepEqual.
1. is caused by the cloning operation itself bombing out on cyclic structures, and
2. is caused by the lack of a check for primitive arguments before copying - strings are copied as if they were arrays. The comments at the head of objectValues are clear about the latter restriction, but this contract limitation is not clear when passed up to propEqual itself.

I enclose a screenshot of the results of comparing two string using propEqual. deepEqual by contrast shows a suitable result.
propEqualStrings

### PR DESCRIPTION
**Pull request by BraulioVM**
Made this pull request to fix issue #411 

The problem within this issue was that propEqual was using the utility function objectValues, which does not check against circular structures, and then compared the result of both objectValues with a normal QUnit.equiv. As this approach does not work, I decided to split QUnit.equiv into  two separate functions:
- `QUnit.equiv.deep`: which performs the old QUnit.equiv computations
- `QUnit.equiv.props`: which compares object properties, ignoring their constructor. This function does check against circular structures and allows propEqual to work as it should. If this function is passed a non-object parameter, a `QUnit.equiv.deep` will be called.

I also created some tests for this new function (mostly copied them from deepEqual.js tests).

Fixes #411

### CODE REVIEW

**Comment by BraulioVM** [comment]
I have signed the CLA several times now, but don't know why it doesn't seem so to jquerybot

File: Gruntfile.js [hunk 1]
```diff
@@ -60,7 +60,8 @@ grunt.initConfig({
 		},
 		all: [
 			"<%= jshint.all %>",
-			"!test/deepEqual.js"
+			"!test/deepEqual.js",
+			"!test/propEqual.js"
```
Comment (jzaefferer) [inline Gruntfile.js]: Let's not add more exceptions. We should fix any issues in new files.

File: src/assert.js [hunk 2]
```diff
@@ -87,23 +87,19 @@ QUnit.assert = Assert.prototype = {
...
 	},
 
 	deepEqual: function( actual, expected, message ) {
-		this.push( QUnit.equiv( actual, expected ), actual, expected, message );
+		this.push( QUnit.equiv.deep( actual, expected ), actual, expected, message );
 	},
 
 	notDeepEqual: function( actual, expected, message ) {
-		this.push( !QUnit.equiv( actual, expected ), actual, expected, message );
+		this.push( !QUnit.equiv.deep( actual, expected ), actual, expected, message );
```
Comment (jzaefferer) [inline src/assert.js]: Let's keep `QUnit.equiv` backwards-compatible. You can still add the `props` property to `QUnit.equiv`. Should remove the need to update the `test/deepEqual.js` tests as well.

File: Gruntfile.js (hunk 1 above)
Comment (BraulioVM) [inline Gruntfile.js]: I added test/propEqual.js to the exceptions because the tests in both files are almost the same

File: src/core.js [hunk 3]
```diff
@@ -74,6 +74,24 @@ var QUnit,
...
 		}
 		return vals;
+	},
+	/**
+	 * Gets the keys of an object which are both enumerable and own
+	 *
+	 * @param {Object} obj
+	 * @return {Array} Array with the enumerable and own keys of the object given
+	 */
+	ownObjectKeys = function( obj ) {
```
Comment (jzaefferer) [inline src/core.js]: This is used only in `src/equiv.js`. Let's add it there. We want to reduce circular dependencies between QUnit modules, should add even more.

File: src/assert.js [hunk 4]
```diff
@@ -87,23 +87,19 @@ QUnit.assert = Assert.prototype = {
 	},
 
 	propEqual: function( actual, expected, message ) {
-		actual = objectValues( actual );
-		expected = objectValues( expected );
-		this.push( QUnit.equiv( actual, expected ), actual, expected, message );
+		this.push( QUnit.equiv.props( actual, expected ), actual, expected, message );
```
Comment (jzaefferer) [inline src/assert.js]: Since this doesn't use `objectValues` anymore, and it isn't used anywhere else, we should remove that function.

**Comment by jzaefferer** [comment]
It looks like `propEqual` now replicates a lot of the `deepEqual` implementation. Maybe we're on the wrong track here. Maybe we should drop `propEqual` and remove the prototype check in `deepEqual`. That's what the io.js project went for their `assert` module: https://github.com/iojs/io.js/pull/621

File: Gruntfile.js (hunk 1 above)
Comment (jzaefferer) [inline Gruntfile.js]: Right, but we shouldn't duplicate crappy code.

**Comment by jzaefferer** [comment]
Not sure why the CLA check is failing, as far as I can tell your signature is fine. You can ignore that, I've pinged someone about that.

**Comment by BraulioVM** [comment]
As you said, I already removed ````objectValues```` and put ````ownObjectKeys```` in equiv.js. I also made ````QUnit.equiv```` backwards compatible. I didn't fix the formatting issues in test/propEqual.js as that could take a really long long time.

Talking about https://github.com/jquery/qunit/pull/776#issuecomment-78273500, I think that when comparing objects, what most people actually want is comparing like propEqual does, but can't really know for sure.

File: src/equiv.js [hunk 5]
```diff
@@ -212,8 +236,75 @@ QUnit.equiv = (function() {
...
+				callers.pop(); // unstack, we are done
+
+				// ensure both objects have the same properties
+				return eq && deepEquiv(aProperties.sort(), bProperties.sort());
+			}
+		}( args[ 0 ], args[ 1 ] ) ) &&
+			propEquiv.apply(this, args.splice( 1, args.length - 1 ) ) );
+
+	},
+	innerEquiv = deepEquiv; // default equivalence check
```
Comment (jzaefferer) [inline src/equiv.js]: Might as well drop `innerEquiv` and assign `deepEquiv.props` directly, then return that.

File: test/deepEqual.js [hunk 6]
```diff
@@ -1,4 +1,4 @@
-QUnit.module( "equiv" );
+QUnit.module( "deepEquiv" );
```
Comment (jzaefferer) [inline test/deepEqual.js]: Let's revert this - there is no exported method called "deepEquiv".

File: test/propEqual.js [hunk 7]
```diff
@@ -0,0 +1,708 @@
+QUnit.module( "propEquiv" );
```
Comment (jzaefferer) [inline test/propEqual.js]: "QUnit.equiv.props"

**Comment by jzaefferer** [comment]
@Krinkle @JamesMGreene @leobalter any input on the discussion about general changes to `deepEqual` (or `propEqual`)?

File: src/equiv.js (hunk 5 above)
Comment (BraulioVM) [inline src/equiv.js]: I'm not sure that would totally make sense. How could ````propEquiv```` be a member of ````deepEquiv````? They are kind of opposite. Something that could possibly be more natural would be to drop ````deepEquiv```` and assign ````innerEquiv```` and ````innerEquiv.props```` directly. The point of this is not having that "linguistic contradiction" with ````propEquiv```` being a subproperty of ````deepEquiv````.

Does that make any sense?

File: test/propEqual.js (hunk 7 above)
Comment (BraulioVM) [inline test/propEqual.js]: Would "equiv.props" be fine? It would be the same pattern used in test/deepEqual.js

File: src/equiv.js (hunk 5 above)
Comment (jzaefferer) [inline src/equiv.js]: Sticking with `innerEquiv` is fine.

**Comment by leobalter** [comment]
I'm sorry I didn't give the proper attention to this PR, should we continue and update?

**Comment by jzaefferer** [comment]
I still think this is a valuable PR and we should try to pick it up and finish it. I just looked over the current stage again and while there's probably some minor details to fix (I noticed a missing end-of-file newline, so there are probably more style issues), overall this should be good to land.

**Comment by jzaefferer** [comment]
Before we put more time into this though, we should figure out what to do about #865. If we contribute our implementation to an external module and make use of that, we can (as easily) share code between `propEqual` and `deepEqual` anymore.

**Comment by leobalter** [comment]
does anyone want to continue the work on this? Should we close this at this point?

**Comment by trentmwillis** [comment]
I think this is still worth pursuing, but unless @BraulioVM wants to take this up again, it'd be best to close and open a new PR.

**Comment by trentmwillis** [comment]
It's been a couple more months, so I'm going to close this and hopefully we can address it in another PR.

**Pull request closed without merging**